Refactor skin construction

peppy opened the pull request:
## Preface

As I continue work on the skin editor, I added support for drag-and-drop sprite importing. While doing so, I noticed that `DefaultSkin`s which have been customised by the user didn't have access to `Sample`s or `Texture`s because that logic is all local to `LegacySkin`.

This PR attempts to bring some sanity to how skins are constructed, but also makes fixing the above quite trivial (have tested that this [works](https://github.com/ppy/osu/compare/master...peppy:skin-editor-sprites?expand=1) with a few more [hacky test lines](https://github.com/ppy/osu/commit/7b41847c2201fea4e9729deb64f7f1a0dfd90e87) of change).

## Testing

I've been relying strictly on unit tests for testing this. Probaly need some more manual testing, especially around configuration fallbacks. Mainly wanted to get this out to ensure from a code-quality perspective everyone is happy, since this will be the start of more changes in a similar direction.

## Changes

Individual changes are split out into commits, but the important ones are listed below:

### 5c4a74378d Remove `Textures` and `Samples` initialisation to `Skin`

The main goal of this pull request.

### 078288a616 Make "skin.ini" the default skin filename and remove redundant parameters

The majority of usages were specifying `skin.ini` manually, so I believe this makes sense.

### 7a1909bf97 Change parameter order of `LegacySkin` to put `IStorageResourceProvider` first

The optional resource store should not be before the required* resource provider.

* note that the resource provider isn't strictly required because of some tests that don't make it. we probably want to clean that up in the future.

### d0ea1739b4 Remove skin configuration stream logic as it is no longer required

Because the base `Skin` implementation is now able to do realm-based lookups (or in the case of fallback, using the provided `IResourceStore`), providing a stream down the whole stack is not required. A filename is enough. Similar reasoning for 9d3c6ade62 and 6c405f1dee.

### 00aea9bef0 Only use legacy resources lookup for protected (aka default) skin

This is still a bit haphazard, and can likely be improved further in follow-up work (maybe using a `virtual IsDefault` or something?). There's an inline comment further explaining what is going on.

### e1236e07ad Fix extensions not being specified in time for realm file caching

I have follow-up work which would potentially delay the caching operation to the first lookup, or as a separate call but that requires some more restructuring (the realm store needs to receive an `ILive` otherwise any access outside its `ctor` is incorrect).

peppy committed:
Rename realm-backed resource store in preparation for non-legacy usage

peppy committed:
Remove `Textures` and `Samples` initialisation to `Skin`

peppy committed:
Replace duplicate `LegacySkinResourceStore` class with `RealmBackedResourceStore`

peppy committed:
Prefer provided resource store over realm backed to keep tests working

peppy committed:
Make providing a custom `ResourceStore` to `LegacyBeatmapSkin` optional (for tests only)

peppy committed:
Documentation improvements

peppy committed:
Fix extensions not being specified in time for realm file caching

peppy committed:
Leave realm resource store construction to base class

peppy committed:
Only use legacy resources lookup for protected (aka default) skin

peppy committed:
Fix `HitObjectSampleTest` adding null files to realm models

peppy committed:
Fix realm backed store not being initialised for some tests

peppy committed:
Ensure `HitObjectSampleTest`'s `TestWorkingBeatmap` provides the marking resource store correctly

peppy committed:
Remove `storage` override from `LegacyBeatmapSkin`

peppy committed:
Remove unnecessary skin reading hack in `DefaultLegacySkin`

peppy committed:
Remove skin configuration stream logic as it is no longer required

peppy committed:
Change parameter order of `LegacySkin` to put `IStorageResourceProvider` first

The optional resource store should not be before the (basically)
required resource provider.

peppy committed:
Make "skin.ini" the default skin filename and remove redundant parameters

peppy committed:
Merge branch 'master' into skin-fuck

peppy committed:
Make `Textures` and `Samples` read-only

bdach reviewed (commented):
Not going to lie, reviewing this was pretty painful. Initially opened diff, couldn't make much out of it, then switched to per-commit, which was still pretty bad (as some changes from earlier commits were reverted later down the line) but somehow got through it and then opened diff again, which was more palatable the second go-around. But I imagine this likely was even worse to write.

Notes from a brief read-through. There are some issues here but none of it seems major, I don't find any changes here particularly offensive.

bdach commented:
I haven't tested this much yet, but there is a weird issue wherein if you set default legacy skin and open skin editor, all HUD elements will change back to lazer default skin somehow:

https://user-images.githubusercontent.com/20418176/159805128-0da7a85b-ced4-4376-9a5a-2c058f28cd5f.mp4

This isn't the case in `master` so I'm guessing something here must have broken it.

peppy commented on osu.Game/Skinning/Skin.cs:
```diff
@@ -41,16 +52,32 @@ public abstract class Skin : IDisposable, ISkin
-                : skin.ToLiveUnmanaged();
+            if (resources != null)
+            {
+                SkinInfo = resources.RealmAccess != null
```
I moved this in the follow up PR already for safety, but apparently doesn’t cause test failures as you noticed.

peppy commented on osu.Game/Skinning/LegacyBeatmapSkin.cs:
```diff
@@ -20,13 +21,27 @@ public class LegacyBeatmapSkin : LegacySkin
 
+        private static IResourceStore<byte[]> createRealmBackedStore(BeatmapInfo beatmapInfo, [CanBeNull] IStorageResourceProvider resources)
+        {
+            if (resources == null)
+                // should only ever be used in tests.
+                return new ResourceStore<byte[]>();
+
+            return new RealmBackedResourceStore(beatmapInfo.BeatmapSet, resources.Files, new[] { @"ogg" });
```
The main issue is that `Skin` takes a `SkinInfo` and this is resolving from a `BeatmapSetInfo`.

I was also hoping to move it but haven’t found a path there just yet.

peppy commented:
I've found the cause and fixed the above issue, but found some others (changing skin while in mania between default and classic doesn't correctly work, possibly due to the fallback lookup:

https://github.com/ppy/osu/blob/994fb966b6ba27baf7c178a7483ccf353db46fc3/osu.Game/Skinning/SkinManager.cs#L205-L217

peppy commented on osu.Game/Skinning/Skin.cs:
```diff
@@ -41,16 +52,32 @@ public abstract class Skin : IDisposable, ISkin
+        /// </summary>
+        /// <param name="skin">The skin's metadata. Usually a live realm object.</param>
+        /// <param name="resources">Access to game-wide resources.</param>
+        /// <param name="storage">An optional store which will be used for looking up skin resources. If null, one will be created from realm <see cref="IHasRealmFiles"/> pattern.</param>
```
Was trying to say that the `SkinInfo`'s files will be used. Originally I tried to abstract this out further to being a realm model (to try and address this issue you [brought up here](https://github.com/ppy/osu/pull/17416#discussion_r833753917)) but can probably reword for now to be more explicit.

peppy committed:
Fix incorrect HUD component fallback

Legacy skins should now always show the legacy hud components. The
conditional here is no longer valid as fallback lookups happen at a
*skin*-fallback level rather than internal *source*-fallback.

Put another way, `LegacyDefaultSkin` (with user customisations) should
still display the classic HUD components even if a font is not provided,
as that font will be available via the skin lookup hierarchy.

The TODO removed in this commit has been already resolved so this code
is no longer required.

peppy committed:
Fix typo in xmldocs

peppy committed:
Reword `storage` param xmldoc to use stronger and better defined langauge

peppy commented:
Above issues are resolved, should be ready for some further testing. I'd recommend we leave this open at least a day or two more to allow people to manually test against edge cases. We have pretty good coverage for skin lookup tests but I'm still not 100% confident.

smoogipoo commented on osu.Game/Skinning/LegacySkin.cs:
```diff
@@ -385,26 +356,15 @@ public override Drawable GetDrawableComponent(ISkinComponent component)
+                                    new LegacyHealthDisplay(),
+                                    new SongProgress(),
+                                    new BarHitErrorMeter(),
+                                }
```
What's with these changes? Is it because the fallback is now being done?

peppy replied:
Yep, check commit message on https://github.com/ppy/osu/pull/17416/commits/a4d17a915f96c505e518dd261f9b4d1d5d967b7f for in depth explanation (wasn't in opening post because it was found in @bdach  review pass)

smoogipoo reviewed (approved):
I've done some pretty rudimentary testing with this and it seems fine.
